src/on_shutdown.c

```diff
@@ -0,0 +1,182 @@
+	 */
+	struct fiber *on_shutdown_fiber;
+	/*
+	 * Main loop
+	 */
+	struct ev_loop *loop;
+};
+
+/*
+ * Shutdown trigger list, not need mutex, because access is
+ * avaliable only from tx thread
+ */
+static RLIST_HEAD(on_shutdown_trigger_list);
+
+static void
+on_shutdown_fiber_wakeup(MAYBE_UNUSED struct ev_loop *loop, ev_async *ev, MAYBE_UNUSED int revents)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger = ev->data;
+	fiber_wakeup(on_shutdown_trigger->on_shutdown_fiber);
+}
+
+static int
+create_new_on_shutdown_trigger(void *arg, void (*handler)(void *), void **opaque)
+{
+	struct on_shutdown_trigger *trigger = (struct on_shutdown_trigger *)
+		malloc(sizeof(struct on_shutdown_trigger));
+	if (trigger == NULL)
+		return -1;
+	trigger->arg = arg;
+	trigger->handler = handler;
+	ev_async_init(&trigger->on_shutdown_async, on_shutdown_fiber_wakeup);
+	trigger->on_shutdown_async.data = trigger;
+	*opaque = trigger;
+	trigger->loop = loop();
+	rlist_add_entry(&on_shutdown_trigger_list, trigger, entry);
+	return 0;
+}
+
+USED int
+on_shutdown_register(void *arg, void (*new_hadler)(void *), void (*old_handler)(void *),
+	void **opaque)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger, *tmp;
+	if (old_handler == NULL) {
+		if (new_hadler != NULL) {
+			return create_new_on_shutdown_trigger(arg, new_hadler, opaque);
+		} else {
+			/*
+			 * Invalid function params, old_handler or new_hadler
+			 * must be set
+			 */
+			return -1;
+		}
+	}
+
+	rlist_foreach_entry_safe(on_shutdown_trigger, &on_shutdown_trigger_list, entry, tmp) {
+		if (on_shutdown_trigger->handler == old_handler) {
+			if (new_hadler != NULL) {
+				/*
+				 * Change on_shutdown trigger handler, and arg
+				 */
+				on_shutdown_trigger->handler = new_hadler;
+				on_shutdown_trigger->arg = arg;
+			} else {
+				/*
+				 * In case new_handler == NULL
+				 * Remove old on_shutdown trigger and destroy it
+				 */
+				*opaque = NULL;
+				rlist_del_entry(on_shutdown_trigger, entry);
+				free(on_shutdown_trigger);
+			}
+			return 0;
+		}
+	}
+	if (new_hadler) {
+		/*
+		 * If we not find on_shutdown trigger but new handler is set
+		 * create new on_shutdown trigger
+		 */
+		return create_new_on_shutdown_trigger(arg, new_hadler, opaque);
+	}
+	/*
+	 * Here we are in case when we not find on_shutdown trigger,
+	 * which we want to destroy return -1.
+	 */
+	return -1;
+}
+
+static int
+on_shutdown_f(va_list ap)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger = va_arg(ap, struct on_shutdown_trigger *);
+	on_shutdown_trigger->handler(on_shutdown_trigger->arg);
+	return 0;
+}
+
+/*
+ * Common function called all module shutdown triggers
+ */
+int
+on_shutdown_trigger_common(MAYBE_UNUSED struct trigger *trigger, MAYBE_UNUSED void *event)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger, *tmp;
+	rlist_foreach_entry_safe(on_shutdown_trigger, &on_shutdown_trigger_list, entry, tmp) {
+		on_shutdown_trigger->on_shutdown_fiber = fiber_new("shutdown", on_shutdown_f);
```

Comment:
I'd mangle name of the fiber so that tell one fiber from another.

Reply:
fiber_new() sets diag message. I guess proper way to handle it is:
diag_log();
diag_clear(diag_get());

Reply:
fix it trigger.cc